### ISSUE TITLE
spatial.distance.cdist segfaults for some structured arrays

### ISSUE THREAD

**Issue by jmlarson1** [scipy/scipy#8891]
When I run the following example in scipy 1.1.0, I get a segfault.

### Reproducing code example:
```
import numpy as np
from scipy.spatial.distance import cdist

H = np.zeros(20, dtype=[('x', '<f8', (2,)), ('m', '<i8'), ('a', '<f8'), ('b', '?'), ('c', '?'), ('d', '<f8'), ('e', '<f8'), ('fa', '<f8'), ('g', '<i8'), ('h', '<i8'), ('i', '?'), ('j', '<i8'), ('k', '?'), ('f', '<f8'), ('l', '?')])
np.random.seed(1)
H['x'] = np.random.uniform(0,1,(20,2))
dist_1 = cdist(np.atleast_2d(H['x'][3]), H['x'], 'euclidean')
```
### Error message: 
(if I import and enable faulthandler)
```
$ python example.py 
Fatal Python error: Segmentation fault

Current thread 0x00007fc9f8280740 <python> (most recent call first):
  File "/homes/jlarson/.local/lib/python2.7/site-packages/scipy/spatial/distance.py", line 2660 in cdist
  File "example.py", line 9 in <module>
Segmentation fault (core dumped)
```

### Scipy/Numpy/Python version information:
```
>>> import sys, scipy, numpy; print(scipy.__version__, numpy.__version__, sys.version_info)
('1.1.0', '1.14.3', sys.version_info(major=2, minor=7, micro=6, releaselevel='final', serial=0))
```

### Further details:
- The segfault does not occur with scipy 1.0.0
- The segfault does not occur if the number of field in the structured array is reduced.

**Comment by rgommers** [comment]
Hmm, I cannot reproduce that. How did you install `scipy`?

**Comment by jmlarson1** [comment]
I installed scipy using `pip install scipy --user -U`

**Comment by rgommers** [comment]
Okay, it's the Python 2.7 manylinux wheel then. I can't test that here. Can anyone reproduce this?

**Comment by jmlarson1** [comment]
My scipy config is below:
```
>>> sp.__config__.show()
lapack_opt_info:
    libraries = ['lapack', 'f77blas', 'cblas', 'atlas', 'f77blas', 'cblas']
    library_dirs = ['/usr/lib/atlas-base/atlas', '/usr/lib/atlas-base']
    define_macros = [('ATLAS_INFO', '"\\"3.10.1\\""')]
    language = f77
    include_dirs = ['/usr/include/atlas']
openblas_lapack_info:
  NOT AVAILABLE
atlas_3_10_blas_threads_info:
  NOT AVAILABLE
atlas_threads_info:
  NOT AVAILABLE
openblas_clapack_info:
  NOT AVAILABLE
atlas_3_10_threads_info:
  NOT AVAILABLE
atlas_blas_info:
    libraries = ['f77blas', 'cblas', 'atlas', 'f77blas', 'cblas']
    library_dirs = ['/usr/lib/atlas-base']
    define_macros = [('HAVE_CBLAS', None), ('ATLAS_INFO', '"\\"3.10.1\\""')]
    language = c
    include_dirs = ['/usr/include/atlas']
atlas_3_10_blas_info:
  NOT AVAILABLE
atlas_blas_threads_info:
  NOT AVAILABLE
openblas_info:
  NOT AVAILABLE
blas_mkl_info:
  NOT AVAILABLE
blas_opt_info:
    libraries = ['f77blas', 'cblas', 'atlas', 'f77blas', 'cblas']
    library_dirs = ['/usr/lib/atlas-base']
    define_macros = [('HAVE_CBLAS', None), ('ATLAS_INFO', '"\\"3.10.1\\""')]
    language = c
    include_dirs = ['/usr/include/atlas']
blis_info:
  NOT AVAILABLE
atlas_info:
    libraries = ['lapack', 'f77blas', 'cblas', 'atlas', 'f77blas', 'cblas']
    library_dirs = ['/usr/lib/atlas-base/atlas', '/usr/lib/atlas-base']
    define_macros = [('ATLAS_INFO', '"\\"3.10.1\\""')]
    language = f77
    include_dirs = ['/usr/include/atlas']
atlas_3_10_info:
  NOT AVAILABLE
lapack_mkl_info:
  NOT AVAILABLE
```

**Comment by WarrenWeckesser** [comment]
> The segfault does not occur if the number of field in the structured array is reduced.

A shot in the dark: does it crash if `H` is simplified to the following?

    H = np.zeros(20, dtype=[('x', '<f8', (2,)), ('b', '?')])

**Comment by rgommers** [comment]
That config looks odd, 1.1.0 should be compiled against OpenBLAS not ATLAS: https://github.com/MacPython/scipy-wheels/blob/v1.1.x/appveyor.yml

**Comment by jmlarson1** [comment]
No, it does not crash with the smaller H. Is there an upper bound on the number of fields in a numpy structured array that spatial.distance supports?

As for the config, I did not specify anything more than `pip install scipy --user -U`. Do I need to specify OpenBLAS somewhere?

**Comment by WarrenWeckesser** [comment]
The function doesn't know that the input arrays are from a structured array.  They are just plain numpy arrays. In the example that crashes, `H['x']` has strides (93, 8), which means the floating point values in that array are not aligned on 8 byte memory boundaries.  But that shouldn't cause a crash; maybe it would just be slower than if the values were nicely aligned.

**Comment by WarrenWeckesser** [comment]
Ignore my "shot in the dark" comments.  Almost the first thing `cdist` does is

    XA = np.asarray(XA, order='c')
    XB = np.asarray(XB, order='c')

which will copy `XA` and `XB` if they are not C-contiguous.

**Comment by pv** [comment]
Misalignment can cause a crash on some non-x86 platforms.

@jmlarson1: what OS and processor is this running on?

**Comment by jmlarson1** [comment]
I am on an x86 platform:
```
$ uname -a
Linux login4 3.13.0-149-generic #199-Ubuntu SMP Thu May 17 10:12:50 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by jmlarson1** [comment]
Similarly, if I use `H['x'][[3]]` instead of `np.atleast_2d(H['x'][3])`, I do not get a segfault.

**Comment by WarrenWeckesser** [comment]
Earlier I said

> Ignore my "shot in the dark" comments. [...] will copy XA and XB if they are not C-contiguous.

But `H['x'][3]`, even though it is not aligned, *is* C-contiguous. In that case, both `np.atleast_2d` (in the script) and `np.asarray` (in the function) will pass the argument through without copying.

If the value is copied (either by using the `.copy()` method or by using `H['x'][[3]]` instead of `np.atleast_2d(H['x'][3])`, the copy will be aligned.  So alignment *could* be the source of the problem.

I don't have an installation where I can reproduce the crash, so I can't investigate further at the moment.

**Comment by ev-br** [comment]
Tried and failed to repro on 
```
$ uname -a
Linux duneyrr 4.13.0-39-generic #44~16.04.1-Ubuntu SMP Thu Apr 5 16:43:10 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

in a fresh virtualenv with python 2.7.12, numpy 1.14.3 and scipy 1.1.0 (manylinux wheel). OP example does not crash.

**Comment by pv** [comment]
Misaligned dereference iirc is undefined behavior in C, so I guess the
compiler may have emitted SSE or other code launching unicorns when
alignment is bad.

So it needs fixing in any case, asarray(order="c") should be replaced
with something that ensures NPY_BEHAVED.

**Comment by mgiraldo** [comment]
is this resolved? i am also getting crashes (even computer restarts) for a large-ish 2d-array (300k points). macOS 10.15.4 (Catalina) Python 3.6.10.

```
$ uname -a
Darwin home 19.4.0 Darwin Kernel Version 19.4.0: Wed Mar  4 22:28:40 PST 2020; root:xnu-6153.101.6~15/RELEASE_X86_64 x86_64
```

i'm trying to fit 300k features into a 576x576 grid. using pyenv and jupyter notebook.

```python
>>> print(grid.shape)
(331776, 2)
>>> print(features.shape)
(331368, 2)
>>> cost = cdist(grid, features, 'sqeuclidean')
CRASH (computer restarted)
```

**Comment by mgiraldo** [comment]
i was using [Quasimondo/RasterFairy](https://github.com/Quasimondo/RasterFairy) to convert UMAPs to grid and it worked with [every one of my datasets](http://dxlab-fellowship-2019.s3-website-ap-southeast-2.amazonaws.com/viewer/) (even one with 550k points) but it was hanging on one of them, possibly due to Python2/3 differences. so I gave [kylemcdonald/CloudToGrid](https://github.com/kylemcdonald/CloudToGrid) a try, which uses `cdist`. i didn't try this approach with the other 20 datasets.
